GHC JS browser introduction blog post/tutorial (#24)
input-output-hk/engineering: blog/2023-01-18-javascript-browser-tutorial.md

---
slug: 2023-01-18-javascript-browser-tutorial

> title: JavaScript backend merged into GHC

Reviewer: Wrong title

Author: Fixed to `Using GHC's JavaScript Backend in the Browser`

> authors: [sylvain, doyougnu, luite, josh, moritz]

Reviewer: ```suggestion
authors: [sylvain, doyougnu, luite, josh]
```

tags: [ghc, javascript, cross-compilation]
---

# Using GHC's JavaScript Backend in the Browser

In the previous
[post](https://engineering.iog.io/2022-12-13-ghc-js-backend-merged) we
introduced GHC's new JavaScript backend, which allows the compilation of Haskell
code into JavaScript. This is the first tutorial in a new series about the
JavaScript backend. We plan to write more of those in the coming weeks and
months as we add new features (e.g. support for "foreign exports" that will
allow JavaScript code to call into Haskell code, support for Template Haskell,
etc.). For now it relies on our "insider" knowledge (e.g. how the FFI works)
that isn't well documented elsewhere. We do plan to add a chapter about the
JavaScript backend in GHC's user guide, but for now your best chance is to look
at GHCJS's documentation or at the source code. In this post, we'll build GHC as
a JavaScript cross-compiler and run a trivial Haskell program in the browser.

> 

Reviewer: Here we should add a warning about the JS backend being a technology preview to lower expectations (cf discussions with David, GHC HQ, etc.)

  setInnerHtml :: CString -> IO ()

circle :: String
circle = "<svg width=300 height=300><circle cx=50% cy=50% r=50%></circle></svg>"

main :: IO ()
main = withCString circle setInnerHtml
```

Notice that we've encountered a Haskell feature that's only available in the
JavaScript backend: JavaScript foreign imports. This feature allows our Haskell
program to call JavaScript functions. In our example we use this feature to call
a JavaScript [arrow
function](https://262.ecma-international.org/13.0/#prod-ArrowFunction) that
updates the `body` of the page with our HTML snippet containing a drawing of a
circle. Alternatively, we could have set the foreign import to a function symbol
like so:

```
foreign import javascript "setInnerHTML"
  setInnerHtml :: CString -> IO ()
```

where `setInnerHTML` is defined in a `.js` file that is then loaded
by passing the JavaScript file to GHC along with the Haskell sources.

Next, we can compile our program to JavaScript, again with our built GHC:

```
ghc-js HelloBrowser.hs
```

Or `ghc-js HelloBrowser.hs foo.js` if `setInnerHTML` is defined in `foo.js`.

> Recall the `index.html` file inside the HelloBrowser.jsexe directory. This HTML

Reviewer: ```suggestion
Recall the `index.html` file inside the `HelloBrowser.jsexe` directory. This HTML
```

file has our compiled JavaScript already included, so if you open it in your
browser, you'll find it loads our SVG circle in the top-left of the page!

![Example webpage screenshot](/img/browser-screenshot.png)

> It's also possible to use our program with existing HTML. In `index.html`, you'll find the line:

Reviewer: ```suggestion
`index.html` contains the minimal HTML code required to load the generated JavaScript code. It simply loads the `all.js` file mentioned above with the following `script` tag that you can reuse in your own HTML files:
```

```html
<script language="javascript" src="all.js" defer></script>
```

> This references the `all.js` file that we described in the first example. So, if we had a HTML document with content, and we wanted to modify it via Haskell, we'd just have to include our program with the `script` tag! 

Reviewer: ```suggestion

As the JS backend still lacks support for some FFI features (foreign exports, foreign "wrapper" imports...), JavaScript codes can't easily interact with Haskell codes. It reduces the amount of advanced/interesting examples we can present for now. We'll publish new blog posts illustrating these features when they will be implemented.
```

authors: [sylvain, doyougnu, luite, josh]
tags: [ghc, javascript, cross-compilation]
---

# Using GHC's JavaScript Backend in the Browser

In a previous
[post](https://engineering.iog.io/2022-12-13-ghc-js-backend-merged) we
introduced GHC's new JavaScript backend, which allows the compilation of Haskell
code into JavaScript. This is the first tutorial in a new series about the
JavaScript backend. We plan to write more of those in the coming weeks and
months as we add new features (e.g. support for "foreign exports" that will
allow JavaScript code to call into Haskell code, support for Template Haskell,
etc.). For now it relies on our "insider" knowledge (e.g. how the FFI works)
that isn't well documented elsewhere. We do plan to add a chapter about the
JavaScript backend in GHC's user guide, but for now your best chance is to look
at GHCJS's documentation or at the source code. In this post, we'll build GHC as
a JavaScript cross-compiler and run a trivial Haskell program in the browser.

Please note: this is a technology preview of the in-development JavaScript backend
for GHC. Not all Haskell features are implemented, and bugs are expected. It is

> currently not possible for JavaScript code to call into Haskell code ("foreign

Reviewer 2: I'd say that it's rather complicated, rather than impossible (you can create callbacks and hand them over to JS. This is how `mkTx` worked, a small `main` that just creates all callbacks for the library to build the library API)

Author: I think calling it impossible is consistent with the messaging that it's a technology preview, even if there's technically a work-around.

Could we get @hsyl20 and @doyougnu's thoughts on this?

Reviewer: ```suggestion
currently rather complicated for JavaScript code to call into Haskell code ("foreign
```

Reviewer: We can soften the wording though even if we don't explain more.

exports" aren't implemented). GHC isn't a multi-target compiler yet, so a GHC executable
built for a native platform (Linux/x86-64, Windows/x86-64, Darwin/AArch64...) as currently distributed (via ghcup, Stack, binary distributions, etc.) won't be able to produce JavaScript. Official prebuilt binary distributions are likely to remain
unavailable until GHC gains multi-target support - requiring the JavaScript backend
to be built from source even after the backend matures.
That's why we start this post with the required steps to build yourself
a GHC compiler capable of producing JavaScript.

## Building GHC as a Cross Compiler to JavaScript

### Installing Dependencies

First we need to install all the typical dependencies for GHC plus `Emscripten`,
so our final list is:

* GHC version 9.2 or later
* Cabal
* Alex
* Happy
* Emscripten to configure with

> * (Optional) NodeJS to run JavaScript locally

Reviewer 2: Emscripten comes with nodejs and adds it to the `PATH` when running `emsdk_env.sh`

Author: IMO it's better to decouple the node installation here. It allows us to mention why it's needed, and lets it be used from a secondary terminal that hasn't `source`d `emsdk_env.sh`.

Let's take these in order, a standard GHC distribution with Cabal is needed so we can boot our new compiler.
We recommend using  GHCUP
([https://www.haskell.org/ghcup/install/](https://www.haskell.org/ghcup/install/)),
or your system's package manager to install the this. 

We need Alex and Happy to build GHC, these can be installed through Cabal:

```
cabal install alex happy -j
```

We need Emscripten during the `configure` step of the build. `Emscripten` should be available in most package managers, but you can also build and install it from source:

```
git clone https://github.com/emscripten-core/emsdk.git
cd emsdk
./emsdk install latest
./emsdk activate latest
source ./emsdk_env.sh
```

After installing Emscripten, `emconfigure` should be available on your system
path. Use `which emconfigure` to check that it is on your `$PATH`. If you built
from source, then the output should point to a location within the emsdk git
project like so:

```
$ which emconfigure

> /path/to/emsdk/upstream/escripten/emconfigure

Reviewer 2: typo, `escripten` -> `emscripten`

Author: typo fixed

```

For more detailed installation instructions, see [https://emscripten.org/docs/getting_started/downloads.html](https://emscripten.org/docs/getting_started/downloads.html). 

> That's all we need to build GHC as a cross compiler. NodeJS can be installed via your system's package manager if you want to run the JavaScript programs locally. We'll assume it's in your `$PATH` for the rest of the blog post.

Reviewer 2: Should already be in `$PATH` after running `emsdk_env.sh`

updates the `body` of the page with our HTML snippet containing a drawing of a
circle. Alternatively, we could have set the foreign import to a function symbol
like so:

```
foreign import javascript "setInnerHTML"
  setInnerHtml :: CString -> IO ()
```

where `setInnerHTML` is defined in a `.js` file that is then loaded
by passing the JavaScript file to GHC along with the Haskell sources.

Next, we can compile our program to JavaScript, again with our built GHC:

```
ghc-js HelloBrowser.hs
```

Or `ghc-js HelloBrowser.hs foo.js` if `setInnerHTML` is defined in `foo.js`.

Recall the `index.html` file inside the `HelloBrowser.jsexe` directory. This HTML
file has our compiled JavaScript already included, so if you open it in your
browser, you'll find it loads our SVG circle in the top-left of the page!

![Example webpage screenshot](/img/browser-screenshot.png)

`index.html` contains the minimal HTML code required to load the generated JavaScript code. It simply loads the `all.js` file mentioned above with the following `script` tag that you can reuse in your own HTML files:
```html
<script language="javascript" src="all.js" defer></script>
```

> As the JS backend still lacks support for some FFI features (foreign exports, foreign "wrapper" imports...), JavaScript codes can't easily interact with Haskell codes. It reduces the amount of advanced/interesting examples we can present for now. We'll publish new blog posts illustrating these features when they will be implemented.

Reviewer 2: Not sure if the `createCallback` API has been merged, if it's the case then we might link to it for people who are a bit more adventurous. Now it looks like it's more limited than it actually is.

Reviewer: I don't know either. Let's keep this for a future blog post IMHO
